### PR TITLE
feat: prevent category combo category changes if data exists [DHIS2-1786]

### PR DESCRIPTION
**Pull request by jbee**
### Summary
Adds a validation to the category combo endpoint to prevent changing the categories of the combo if data exist.

Depends on #11618 to make the added validation be effective for all ways of updating an object. 

### Automatic Testing
Added a new integration test to check basic functionality of the service method that checks for existing data values.

### Manual Testing
* find or create a category combo where data exist
* try to add or remove a category to/from the combo and check that the attempt results in an error
* find or create a category combo where no data exists
* try to add or remove a category to/from the combo and check that it is permitted
* check the above both for data values where the combo is used of category option combo and/or as attribute option combo

### CODE REVIEW

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/datavalue/DataValueService.java [hunk 1]
```diff
@@ -205,4 +206,12 @@ void validate( DataExportParams params )
...
      */
     int getDataValueCountLastUpdatedBetween( Date startDate, Date endDate, boolean includeDeleted );
+
+    /**
+     * Checks if any data values exist for the provided {@link CategoryCombo}
+     *
+     * @param combo the combo to check
+     * @return true, if any value exist, otherwise false
+     */
+    boolean existsAnyValue( CategoryCombo combo );
```
Comment (larshelge) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/datavalue/DataValueService.java:216]: Perhaps a slightly more natural method name `dataValueExists( CategoryCombo category Combo )` ?

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/datavalue/DataValueService.java [hunk 2]
```diff
@@ -205,4 +206,12 @@ void validate( DataExportParams params )
      * @return the number of DataValues.
      */
     int getDataValueCountLastUpdatedBetween( Date startDate, Date endDate, boolean includeDeleted );
+
+    /**
+     * Checks if any data values exist for the provided {@link CategoryCombo}
```
Comment (larshelge) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/datavalue/DataValueService.java:211]: Very minor, end sentences with `.` in Javadoc.

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/datavalue/DataValueStore.java [hunk 3]
```diff
@@ -157,4 +158,12 @@ DataValue getDataValue( DataElement dataElement, Period period, OrganisationUnit
...
      */
     int getDataValueCountLastUpdatedBetween( Date startDate, Date endDate, boolean includeDeleted );
+
+    /**
+     * Checks if any data values exist for the provided {@link CategoryCombo}
+     *
+     * @param combo the combo to check
+     * @return true, if any value exist, otherwise false
+     */
+    boolean existsAnyValue( CategoryCombo combo );
```
Comment (larshelge) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/datavalue/DataValueStore.java:168]: Same as above.

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/feedback/ErrorCode.java [hunk 4]
```diff
@@ -59,6 +59,7 @@
         "with the separator character in one of its codes: `{2}`" ),
     E1118( "Option set `{0}` of value type multi-text cannot have option codes with the separator character: `{1}`" ),
     E1119( "{0} already exists: `{1}`" ),
+    E1120( "Change cannot be applied as it would make existing data values inaccessible." ),
```
Comment (larshelge) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/feedback/ErrorCode.java:62]: Perhaps use `Update` instead of `Change`. Do not end error codes with `.`.

**Review by larshelge — APPROVED** [review]

**Comment by larshelge** [comment]
Nice work!

**Review by mortenoh — APPROVED** [review]

**Pull request merged**